python/taichi/ad/_ad.py

```diff
@@ -4,19 +4,31 @@
 gradient computation task.
 """
 import warnings
+from copy import deepcopy
 from functools import reduce
 
+import matplotlib.pyplot as plt
 import numpy as np
+import taichi.types.primitive_types as types
 from taichi.lang import impl
 from taichi.lang.enums import SNodeGradType
+from taichi.lang.expr import Expr
 from taichi.lang.field import ScalarField
+from taichi.lang.kernel_impl import kernel
 from taichi.lang.snode import SNode
+from taichi.types import ndarray, template
 
 from taichi import _snode
 
 
 class Tape:
-    def __init__(self, loss=None, clear_gradients=True, validation=False):
+    def __init__(self,
+                 loss=None,
+                 clear_gradients=True,
+                 validation=False,
+                 grad_check=None,
+                 watch=None,
+                 eps_range=2.**np.arange(-3, -30, -1).astype(np.float64)):
```

Comment:
I think it is not necessary to expose the `eps_range` since it is too tricky for users to tuning this parameter.

Reply:
I see